### DISCUSSION TITLE
How s-tui compute Power Consumption of package, core and other stuff ?

### DISCUSSION THREAD

**Discussion by kuolemaaa** [amanusk/s-tui#216] [Q&A]
Is there a resource on the sysfs that tells the instantaneous power ?
Is there a way to know where it get this information ?
I'm not talking about a laptop, notebook or something with batteries. I'm talking about on a classic desktop computer (intel)

**Comment by amanusk** [comment]
You are welcome to inspect the code :) 
The power is read from dedicated Intel and AMD registers. On Intel this is called Intel RAPL. It is quite well documented.

RAPL does not rely on battery drainage to estimate power.

It used to be possible to read these registers w/o root access, but this was changed in more recent Linux kernels, as it exposes a side channel security risk.

**Reply by kuolemaaa** [reply]
> On Intel this is called Intel RAPL

Is that the thing on `/sys/class/powercap/intel-rapl` ? I found measurements of energy consumed since I don't know. I didn't find any instant power absorption 

I gathered all my courage and inspected the code and it seems it was as I feared: difference between two energy reading divided by time interval.

>as it exposes a side channel security risk.

very cool

**Reply by berglh** [reply]
@amanusk 

I just built a little home server build with an AMD 9600X and was surprised to see these pacakge and (single) core power values. Seems like kernel 6.14 seems to expose the AMD energy core/package registers via sysfs Intel RAPL as well, at least on my Asus MB.

Looking at the source code, it seems to default to Intel RAPL if it exists first, which I guess makes sense to exit quickly for the first possible match, however, is this less accurate than the AMDRaplMsrReader for AMD?

At the end of the day, I'm more interested in the total package power, which Intel RAPL seems to report on my Zen 5/MB combination - so the default suits me fine, but it seems like per core power usage is also available in AMDRaplMsrReader.

Anyway, great tool, enjoying it.

I've also been working on GTK4 UI tool in my spare time for monitoring Zen, but not working on it for a while. I was using it to read data via the `ryzen_smu` module. I wasn't aware that you can access `msr` via the method shown in this class pretty neat stuff, I'll need to see what I can leverage from `/dev/cpu/*/msr` instead of `ryzen_smu`, although it might be a separate set of registers for each specific CPU rather than the System Management Unit, it could still be useful for power reporting :)
